### PR TITLE
Tao 877 student progress indicator

### PR DESCRIPTION
**Pull request by jsconan**
Related to: https://oat-sa.atlassian.net/browse/TAO-877

The progress bar can be displayed using two modes:
- percentage, the classic way with percentage of completed items
- position, which display the position of the currently viewed item within the entire test

The mode is set up from a config entry inside the taoQtiTest extension: `progress-indicator`

As this extension did not have a config previously set, such a config has been added. And this new config need to be handled by updater on already installed systems. Unfortunately the extension mainly uses the old class path system, and the added Updater relies on namespace. To be able to properly update the extension a `composer update` needs to be executed, and this can be done only from the `develop` branch...

So to test the feature without running an update, a change inside the code is needed. By example by forcing directly the mode in the code. To do so just set this code into the `views/js/controller/runtime/testRunner.js` file at line 20:

```javascript
var progressIndicator = 'position'; // or 'percentage'
```

### CODE REVIEW

File: views/js/controller/runtime/testRunner.js [hunk 1]
```diff
@@ -7,6 +7,51 @@ define(['jquery', 'lodash', 'spin', 'serviceApi/ServiceApi', 'serviceApi/UserInf
...
+            /**
+             * Updates the progress bar displaying the position
+             * @param {Object} assessmentTestContext The progression context
+             */
+            positionProgression: function(assessmentTestContext) {
+                var total = assessmentTestContext.numberItems;
+                var position = assessmentTestContext.itemPosition + 1;
+                return {
+                    ratio : Math.floor(position / total * 100),
+                    label : __('Item %X on %Y').replace('%X', position).replace('%Y', total)
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: the function  `__` accept basic formatting :  `__('Item %d on %d', position, total)` (syntax to be verified)

File: views/js/controller/runtime/testRunner.js [hunk 2]
```diff
@@ -7,6 +7,51 @@ define(['jquery', 'lodash', 'spin', 'serviceApi/ServiceApi', 'serviceApi/UserInf
...
+                    label : __('Test completed at %d%%').replace('%d', ratio).replace('%%', '%')
+                };
+            },
+
+            /**
+             * Updates the progress bar displaying the position
+             * @param {Object} assessmentTestContext The progression context
+             */
+            positionProgression: function(assessmentTestContext) {
+                var total = assessmentTestContext.numberItems;
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: check the value (to prevent a /0)

File: views/js/controller/runtime/testRunner.js [hunk 3]
```diff
@@ -7,6 +7,51 @@ define(['jquery', 'lodash', 'spin', 'serviceApi/ServiceApi', 'serviceApi/UserInf
...
+        var progressUpdaters = {
+            /**
+             * Updates the progress bar
+             * @param {Object} assessmentTestContext The progression context
+             */
+            update: function(assessmentTestContext) {
+                var progressIndicator = assessmentTestContext.progressIndicator || 'percentage';
+                var progressIndicatorMethod = progressIndicator + 'Progression';
+                var getProgression = this[progressIndicatorMethod] || this.percentageProgression;
+                var progression = getProgression(assessmentTestContext);
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: check `getProgression`

**Comment by jsconan** [comment]
Use parametric translation for the progress message
Checked getProgression method
Prevent division by zero

File: scripts/update/Updater.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @return string $versionUpdatedTo
+     */
+    public function update($initialVersion) {
+
+        $currentVersion = $initialVersion;
+        
+        // add testrunner config
+        if ($currentVersion == '2.6') {
+
+            common_ext_ExtensionsManager::singleton()->getExtensionById('taoQtiTest')->setConfig('testRunner', array(
```
Comment (llecaque) [inline scripts/update/Updater.php]: Please be more cautious when testing update script, 

taoQtiTest requires update from 2.6 to 2.6.1
  Running oat\taoQtiTest\scripts\update\Updater
PHP Fatal error:  Class 'oat\taoQtiTest\scripts\update\common_ext_ExtensionsManager' not found in /var/www/taopootle/tao30/taoQtiTest/scripts/update/Updater.php on line 40

Comment (krampstudio) [inline scripts/update/Updater.php]: Merge error...

**Comment by llecaque** [comment]
solved in https://github.com/oat-sa/extension-tao-testqti/pull/80

**Pull request merged**